[core] feat(TagInput): resizable input that scales as you type

JosephChotard opened the pull request:
#### Checklist

- [N/A] Includes tests
- [N/A] Update documentation

<!-- DO NOT enable CircleCI for your fork. Our build will run when you open this PR. -->

#### Changes proposed in this pull request:

Currently the `TagInput` component (used by `MultiSelect2`) does not expand to accomodate the text the user is typing. This is because `input` elements don't scale with their input. This causes annoying situations where the user cannot see what they are typing when there is space to display it. 
To address this issue, I created a small `ResizeableInput` component (could be included in `TagInput` but this makes the intent clearer). This new component acts like a regular `input` element for every other component but it has some coding magic that causes it to fit the size of it's content. It will never grow bigger than it's closest parent with a `position: relative`. 

#### Reviewers should focus on:

This applies to both `TagInput` and `MultiSelect`
I checked that:
* The input never stretches wider than the parent
* The input scales with it's content
* The input is never smaller than 80px
* The existing props still work (large, disabled, add on blur, ...)

This PR does not introduce changes to the API and is fully backwards compatible.

#### Screenshot

Old Input scaling:
<img width="535" alt="numbers - old" src="https://user-images.githubusercontent.com/33576267/220661307-75a94214-dfa4-482a-ac4b-e3ab02bdf655.png">
New Input scaling:
<img width="411" alt="numbers - new" src="https://user-images.githubusercontent.com/33576267/220661299-dcfe52d9-6f3b-4482-b862-f752f3a4d17e.png">

Old input text being cut off
<img width="514" alt="wrap old" src="https://user-images.githubusercontent.com/33576267/220661322-81e29e1c-02cb-4b46-b7e1-e5ba8bf5b40e.png">

New input text wrapping
<img width="519" alt="wrap new" src="https://user-images.githubusercontent.com/33576267/220661339-27109854-1dcc-4fd9-aeae-e9308db20213.png">

JosephChotard committed:
make input resizeable

JosephChotard committed:
fix: white space no wrap

JosephChotard committed:
fix: preserve spaces

adidahiya commented on packages/core/src/components/tag-input/resizeableInput.tsx:
```diff
@@ -0,0 +1,50 @@
+type ResizeableInputProps = HTMLInputProps;
+export type Ref = HTMLInputElement;
+
+export const ResizeableInput = forwardRef<Ref, ResizeableInputProps>(function ResizeableInput(props, ref) {
```
I believe "resizable" is more correct, at least for US English. We use that elsewhere in Blueprint, so please rename all references to "resizeable" -> "resizable" (including the file name).

JosephChotard replied:
ack

adidahiya commented on packages/core/src/components/tag-input/resizeableInput.tsx:
```diff
@@ -0,0 +1,50 @@
+    const span = useRef<HTMLSpanElement>(null);
+
+    useEffect(() => {
+        if (span.current) {
```
nit: please use explicit boolean coercion

```suggestion
        if (span.current != null) {
```

adidahiya commented on packages/core/src/components/tag-input/resizeableInput.tsx:
```diff
@@ -0,0 +1,50 @@
+
+    const { onChange, ...otherProps } = props;
+
+    const changeHandler: React.ChangeEventHandler<HTMLInputElement> = evt => {
```
nit: use `handle...` naming for event callbacks

```suggestion
    const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = evt => {
```

adidahiya commented on packages/core/src/components/tag-input/resizeableInput.tsx:
```diff
@@ -0,0 +1,50 @@
+            <span
+                id="hide"
+                ref={span}
+                style={{
```
prefer styling with CSS rather than inline styles. you should create a `_resizableInput.scss` file next to this one and create a new class following our existing naming conventions...

```scss
.#{$ns}-resizable-input-span {
  max-height: 0;
  max-width: 100%;
  min-width: 80px;
  position: absolute;
  white-space: nowrap;
  z-index: -100;
}
```

adidahiya commented on packages/core/src/components/tag-input/resizeableInput.tsx:
```diff
@@ -0,0 +1,50 @@
+                    minWidth: "80px",
+                    position: "absolute",
+                    whiteSpace: "nowrap",
+                    zIndex: -100,
```
why is this necessary? can it be simply `-1`? can you leave a code comment?

JosephChotard replied:
whoops, fixing it

adidahiya commented on packages/core/src/components/tag-input/resizeableInput.tsx:
```diff
@@ -0,0 +1,50 @@
+                    zIndex: -100,
+                }}
+            >
+                {/* Need to replace spaces with the html character for them to be preserved */}
```
:+1:

adidahiya reviewed (changes requested):
Nice, this seems to work really well in the [docs preview build](https://output.circle-artifacts.com/output/job/5d65e360-4486-435a-998d-09b9bccb2b61/artifacts/0/packages/docs-app/dist/index.html#core/components/tag-input)!

Looks like the tests are failing because we're using shallow rendering in many of the TagInput tests, and the `<ResizeableInput>` children don't get rendered. We can fix that by switching to `mount()` instead of `shallow()`.

JosephChotard committed:
tests: change simulate to call function directly

JosephChotard commented:
https://github.com/enzymejs/enzyme/issues/1473
https://github.com/enzymejs/enzyme/issues/1081
Changed the tests to use `mount` and modified a few `.simulate` calls as they caused the tests to hang (not sure why), instead I call the prop directly, I don't believe we need to test the event passing as that's React's responsibility.

JosephChotard committed:
rename to resizable

JosephChotard committed:
use explicit boolean coercion

JosephChotard committed:
[refactor] move styles to scss file

JosephChotard committed:
[refactor] rename using handle... syntax

JosephChotard commented:
@adidahiya done all the changes, lmk if there's anything else

adidahiya commented on packages/core/src/common/classes.ts:
```diff
@@ -163,6 +163,8 @@ export const INPUT_GROUP = `${INPUT}-group`;
 export const INPUT_LEFT_CONTAINER = `${INPUT}-left-container`;
 export const INPUT_ACTION = `${INPUT}-action`;
 
+export const INPUT_RESIZABLE_INPUT_SPAN = `${NS}-resizable-input-span`;
```
nit: the variable name should match the class name more closely:

```suggestion
export const RESIZABLE_INPUT_SPAN = `${NS}-resizable-input-span`;
```

adidahiya commented on packages/core/src/components/_index.scss:
```diff
@@ -37,6 +37,7 @@
 @import "tabs/tabs";
 @import "tag/tag";
 @import "tag-input/tag-input";
+@import "tag-input/resizeable-input";
```
missed a file name here:

```suggestion
@import "tag-input/resizable-input";
```

adidahiya commented on packages/core/src/components/tag-input/resizableInput.tsx:
```diff
@@ -0,0 +1,39 @@
+import { Classes, HTMLInputProps } from "../../common";
+
+type ResizableInputProps = HTMLInputProps;
+export type Ref = HTMLInputElement;
```
I don't think this typedef needs to exist. Just inline the type: `forwardRef<HTMLInputElement, ResizableInputProps>`

adidahiya commented on packages/core/src/components/tag-input/resizableInput.tsx:
```diff
@@ -0,0 +1,39 @@
+
+import { Classes, HTMLInputProps } from "../../common";
+
+type ResizableInputProps = HTMLInputProps;
```
nit: export for consistency

```suggestion
export type ResizableInputProps = HTMLInputProps;
```

adidahiya commented on packages/core/test/tag-input/tagInputTests.tsx:
```diff
@@ -32,19 +26,17 @@ import { Button, Classes, Intent, Keys, Tag, TagInput, TagInputProps } from "../
         // check that event handler is proxied
         const fakeEvent = { flag: "yes" };
-        input.simulate("blur", fakeEvent);
+        input.prop("onBlur")?.(fakeEvent as any);
```
I don't love this change but it appears to be the best path forward given the Enzyme issues you just linked. I'm surprised to learn that `.simulate()` is no longer recommended. And at some point I will move away from Enzyme anyway so these tests will get updated...

JosephChotard replied:
agreed

adidahiya reviewed (changes requested):
looking good, just a few more minor things

JosephChotard committed:
refactor: code style

JosephChotard committed:
refactor: rename style file

JosephChotard commented:
@adidahiya thanks for all the feedback, should be addressed now

adidahiya reviewed (approved):
LGTM, thanks for the thorough PR @JosephChotard :+1:

bvandercar-vt commented:
@adidahiya this change has created a bug where the `placeholder` is no longer present for a `MultiSelect2`

nerdstep commented:
Looks like there might be a regression with the input width not expanding to the full width of the component when using the `fill` prop:

![image](https://user-images.githubusercontent.com/535021/221920321-42d3074d-17ad-4fc7-a38a-f949eb142a08.png)

bvandrc commented on packages/core/src/components/tag-input/resizableInput.tsx:
```diff
@@ -0,0 +1,38 @@
+
+    return (
+        <span>
+            <span id="hide" ref={span} className={Classes.RESIZABLE_INPUT_SPAN}>
```
Not sure why this `id` prop is needed / where it's used, but it's causing aXe a11y failures because all `id`s must be **unique**.

Must either remove `id` prop, or use `uniqueId` for it.

bvandrc replied:
PR: https://github.com/palantir/blueprint/pull/5984

adidahiya commented:
@nerdstep thanks for pointing out the regression. Fix here: https://github.com/palantir/blueprint/pull/5985
